### PR TITLE
SAK-46617 Tests & Quizzes: Exceptions have restrictions not present for non-excepted students

### PR DESCRIPTION
**Pull request by ropemar**
https://sakaiproject.atlassian.net/browse/SAK-46617

I've deleted the validation for the due date being greater or equal to the current date and all properties refered to this restriction.

### CODE REVIEW

**Comment by mpellicer** [comment]
Please rebase it @ropemar

**Comment by ottenhoff** [comment]
I'd like to hear from @bjones86 on this

**Comment by bjones86** [comment]
I'm in favour of removing this outright prevention. However, I think it would be better to provide a warning in this scenario, just to avoid any unintentional mistakes.

**Review by bjones86 — CHANGES_REQUESTED** [review]
I would prefer to see this implemented as a warning. The warning should be presented if the due date is in the past, but should be allowed if the instructor wishes to continue.

**Comment by ropemar** [comment]
Changed the PR to convert this validation to a warning instead of a mandatory requisite. In order to make this possible, I had to implement a separate visualization of warnings and error messages, since the settings form only showed error message boxes.

**Comment by bjones86** [comment]
Thanks, @ropemar. Can you add a screenshot of this new functionality to the JIRA ticket?

**Comment by ropemar** [comment]
Yes, of course. You have it right now on the JIRA.

File: samigo/samigo-app/src/java/org/sakaiproject/tool/assessment/ui/bean/author/AssessmentSettingsBean.java [hunk 1]
```diff
@@ -1987,4 +1988,36 @@ private void resetExtendedTime() {
...
+    
+    public List<FacesMessage> getErrorMessages() {
+        List<FacesMessage> list = new ArrayList<>();
+        FacesContext context = FacesContext.getCurrentInstance();
+        for (Iterator<FacesMessage> iter = context.getMessages() ; iter.hasNext(); ) {
+        	FacesMessage message = iter.next();
+        	if (FacesMessage.SEVERITY_WARN.equals(message.getSeverity())) {
+        		list.add(message);
+        	}
+        }
```
Comment (bjones86) [inline samigo/samigo-app/src/java/org/sakaiproject/tool/assessment/ui/bean/author/AssessmentSettingsBean.java:2000]: You could simplify this pattern with a stream operation.

File: samigo/samigo-app/src/java/org/sakaiproject/tool/assessment/ui/bean/author/AssessmentSettingsBean.java [hunk 2]
```diff
@@ -1987,4 +1988,36 @@ private void resetExtendedTime() {
...
+    
+    public List<FacesMessage> getInfoMessages() {
+        List<FacesMessage> list = new ArrayList<>();
+        FacesContext context = FacesContext.getCurrentInstance();
+        for (Iterator<FacesMessage> iter = context.getMessages() ; iter.hasNext(); ) {
+        	FacesMessage message = iter.next();
+        	if (FacesMessage.SEVERITY_INFO.equals(message.getSeverity())) {
+        		list.add(message);
+        	}
+        }
```
Comment (bjones86) [inline samigo/samigo-app/src/java/org/sakaiproject/tool/assessment/ui/bean/author/AssessmentSettingsBean.java:2012]: Same here.

File: samigo/samigo-app/src/java/org/sakaiproject/tool/assessment/ui/bean/author/AssessmentSettingsBean.java [hunk 3]
```diff
@@ -1987,4 +1988,36 @@ private void resetExtendedTime() {
...
+        return list;
+    }
+    
+    public boolean isRenderErrorMessage() {
+        return getErrorMessages().size() > 0;
```
Comment (bjones86) [inline samigo/samigo-app/src/java/org/sakaiproject/tool/assessment/ui/bean/author/AssessmentSettingsBean.java:2017]:
```suggestion
        return !getErrorMessages().isEmpty();
```

File: samigo/samigo-app/src/java/org/sakaiproject/tool/assessment/ui/bean/author/AssessmentSettingsBean.java [hunk 4]
```diff
@@ -1987,4 +1988,36 @@ private void resetExtendedTime() {
...
+    }
+    
+    public boolean isRenderInfoMessage() {
+        return getInfoMessages().size() > 0;
```
Comment (bjones86) [inline samigo/samigo-app/src/java/org/sakaiproject/tool/assessment/ui/bean/author/AssessmentSettingsBean.java:2021]:
```suggestion
        return !getInfoMessages().isEmpty();
```

File: samigo/samigo-app/src/java/org/sakaiproject/tool/assessment/bundle/AssessmentSettingsMessages.properties [hunk 5]
```diff
@@ -377,7 +377,7 @@ extendedTime_entry_table_title=Add/edit an extended time entry.
 # Extended Time Errors
 extended_time_user_and_group_set=A user or group must be selected to add a date and/or time limit exception.
 extended_time_due_earlier_than_available=The Due Date cannot be earlier than the Available Date for the exception set for {0}.
-extended_time_due_in_past=The Due Date cannot be in the past for the exception set for {0}.
+extended_time_due_in_past=The Due Date is in the past for the exception set for {0}.
```
Comment (bjones86) [inline samigo/samigo-app/src/java/org/sakaiproject/tool/assessment/bundle/AssessmentSettingsMessages.properties:380]: This is a good start. I would expand on this text, something like:

```suggestion
extended_time_due_in_past=The Due Date is in the past for the exception set for {0}. Please change the Due Date, or Save again to proceed as is.
```

@fostersdesign, any suggestions here?

**Review by bjones86 — CHANGES_REQUESTED** [review]
Thank you for implementing this as a warning! I just have some very minor comments.

File: samigo/samigo-app/src/java/org/sakaiproject/tool/assessment/ui/bean/author/AssessmentSettingsBean.java (hunk 1 above)
Comment (ropemar) [reply samigo/samigo-app/src/java/org/sakaiproject/tool/assessment/ui/bean/author/AssessmentSettingsBean.java]: Thanks, you're right. It's far more elegant with stream :)

File: samigo/samigo-app/src/java/org/sakaiproject/tool/assessment/ui/bean/author/AssessmentSettingsBean.java (hunk 3 above)
Comment (ropemar) [reply samigo/samigo-app/src/java/org/sakaiproject/tool/assessment/ui/bean/author/AssessmentSettingsBean.java]: Changed, thank you!

File: samigo/samigo-app/src/java/org/sakaiproject/tool/assessment/bundle/AssessmentSettingsMessages.properties (hunk 5 above)
Comment (ropemar) [reply samigo/samigo-app/src/java/org/sakaiproject/tool/assessment/bundle/AssessmentSettingsMessages.properties]: Changed to your suggestion.

**Comment by bjones86** [comment]
Thanks, @ropemar! I'd just like to wait for @fostersdesign to review the message language.

**Comment by fostersdesign** [comment]
The wording looks okay to me. 

The Teaching & Learning group and the Sakai UX group were just talking last week about this convention of warning people of dates in the past and having them click the button again to proceed in Assignments. We felt it was a clunky way to confirm a warning and so we proposed some changes on SAK-46356. That shouldn't block this work from getting in. In the future, the replacement for this convention should probably be standardized across tools though.

**Review by bjones86 — APPROVED** [review]

**Pull request merged**